新規画面 > app wallet 3画面

yuki-yamamura opened the pull request:
- close #42
- close #50 

## 対応したこと

- 既存のコンポーネントを調整
  - カード > 汎用コンポーネントとして不十分だった点を修正
  - テキストフィールド > 最大幅まで伸びる様に修正
- App Wallet ページの追加
  - Transfer
  - Staking
  - Buy

## 対応しないこと

実装が間に合わなかったので、別issueで対応とさせてください 🙏 
- Transferページのタブリンク
- SP スタイル崩れ
- テキストフィールドのアイコン表示 (Slackにて確認中)

## Remarks

以下は伊藤さんと口頭で話して調整しました。
- リンクボタンのborderの色はすべて統一でok
- crypto walletのデザインに合わせてヘッダーはページごとに変更
- タブリンクはFigma通り、Transferページにのみ表示
- Stakingページの "Reference APR" などのテキストは固定でok

## スクリーンショット

### Transfer
<img width="752" alt="スクリーンショット 2024-11-08 21 04 36" src="https://github.com/user-attachments/assets/aaebfde3-26c0-4c35-b3e4-e21d520006e7">

### Staking
<img width="770" alt="スクリーンショット 2024-11-08 21 17 50" src="https://github.com/user-attachments/assets/536d2b1d-cdae-4bc6-a97e-93efecf2b4f6">

### Buy
<img width="748" alt="スクリーンショット 2024-11-08 21 17 29" src="https://github.com/user-attachments/assets/05960fb7-0c59-468f-a1ce-1bab0095e417">

yuki-yamamura committed:
feat: wallet > ルーティングを追加する

yuki-yamamura committed:
refactor: カードコンポーネントのファイル名と記述を他に合わせる

yuki-yamamura committed:
fix: カードコンポーネントを汎用的に使う上で都合が悪い部分を修正する

yuki-yamamura committed:
fix: テキストフィールドを最大幅まで広げる

yuki-yamamura committed:
feat: Wallet > Transferページを追加する

yuki-yamamura committed:
feat: Wallet > Stakingページを追加する

yuki-yamamura committed:
feat: Wallet > Buyページを追加する

yuki-yamamura commented on src/features/app/components/thanks-card.tsx:
```diff
@@ -6,12 +6,13 @@ import type { ReactNode } from "react";
   linkButtons: {
     label: string;
     to: string;
+    isCurrentPage?: boolean;
```
app walletとcrypto walletページでは、カレントページの場合にborderを表示する必要がありました。
その考慮が抜けていたので追加しています。

yuki-yamamura commented on src/features/app/wallet/components/transfer.tsx:
```diff
@@ -0,0 +1,93 @@
+  const user = data.data();
+  const thanks = user?.thanks ?? 0;
+
+  const handleFormSubmit = (e: FormEvent<HTMLButtonElement>) => {
```
フォーム送信の挙動を止められてないかもしれないので後で確認します 🙏

ryosuke-sato-51 replied:
handleFormSubmitをButtonのonAbortイベントで渡していると思うのですが、（onAbortイベントは初めて知りました）これをformのonSubmitに渡すようにしてe.preventDefault()で無効にする感じでいけるかな〜って感じです。参考までに！

```
<form onSubmit={handleFormSubmit}>
```

yuki-yamamura replied:
ありがとうございます 🙇 `handleFormSubmit` を渡す先の要素から属性まですべて間違ってました orz
教えていただいた方法でフォームの送信を止められました！

yuki-yamamura commented on src/features/app/components/thanks-card.tsx:
```diff
@@ -6,12 +6,13 @@ import type { ReactNode } from "react";
     to: string;
+    isCurrentPage?: boolean;
   }[];
+  header?: ReactNode;
```
header必要ない箇所からも利用する必要があったのでオプショナルにして、カードの高さが崩れない様に調整しました。

yuki-yamamura commented:
@ryosuke-sato-51 
レビューをお願いします！
crypto walletで使用するコンポーネントを一部調整させていただきました。
もし問題がなければ、このPRをマージしてから着手していただいた方がいいかもしれません 🙏

ryosuke-sato-51 commented on src/features/app/wallet/components/buy.tsx:
```diff
@@ -0,0 +1,71 @@
+        <LinkButton
+          // TODO: リンク先をStripeの購入ページに変更する
+          to="#"
+          type="button"
```
[nits]
リンクなのでtype="button" は必要なさそう？

ryosuke-sato-51 commented on src/features/app/wallet/components/staking.tsx:
```diff
@@ -0,0 +1,100 @@
+import { TextField } from "@/components/ui/form/text-field";
+import { useGetUser } from "@/features/profile/api";
+import { path } from "@/utils/path";
+import { FormControlLabel, Stack } from "@mui/material";
```
FormControlLabelですが、MUIでフォーム系のコンポーネントはsrc/components/ui/form以下にラッパーコンポーネント作っているので、そこに新規作成した方が良いかもです

yuki-yamamura replied:
@ryosuke-sato-51 
承知しました！
`src/components/ui/form` にコンポーネントを作成して、それを呼び出す様に変更します 🙏 

[question]
FormControlLabelですが、単にMUIのコンポーネントをラップするだけになってしまいました。
このようなケースにおいても、「UIに変更があった際の影響を局所的にできる」といった利点があるためラップした方がよいでしょうか？
`src/components/ui` 以下にファイルをつくる際の基準がわかっていない気がするので、考え方などがあれば教えていただきたいです 🙇

ryosuke-sato-51 replied:
@yuki-yamamura 
そうですね！自分も同じような認識なのですが、今後例えばMUIで破壊的な変更があったりした際に、影響が局所的に出来るイメージです！
ラッパーコンポーネントを作る基準は明確に決めれていないのですが、Stackのようなレイアウト系のコンポーネントや、FormControlのような機能だけ提供するコンポーネント以外は作っておいて良いかな〜という考え方です。

yuki-yamamura replied:
@ryosuke-sato-51 
ありがとございます！プロジェクト内ではこの考え方でやってみます 👍 
> Stackのようなレイアウト系のコンポーネントや、FormControlのような機能だけ提供するコンポーネント以外は作っておいて良いかな〜という考え方です。

ryosuke-sato-51 commented on src/features/app/wallet/constants/index.ts:
```diff
@@ -0,0 +1,14 @@
+const WalletEnum = {
+  APP_WALLET: 1,
+  CRYPTO_WALLET: 2,
+} as const;
+
+const walletOptions = [
+  { value: WalletEnum.APP_WALLET, label: "app wallet" },
+  {
+    value: WalletEnum.CRYPTO_WALLET,
+    label: "crypt wallet",
+  },
+] as const;
```
定数だと思うので命名を大文字のスネークケースにした方が、パッと見で分かりやすいかと思いました。

yuki-yamamura replied:
承知しました！
こちら、アッパースネークケースにするとこんな感じになりました。

[question]
オブジェクトの定数もアッパースネークケースで命名することが多いですか？
(自分の内でこれといったルールが持てておらず... 🙏 )

```ts
const WALLET_MAP = {
  APP_WALLET: 1,
  CRYPTO_WALLET: 2,
} as const;

const WALLET_OPTIONS = [
  { value: WALLET_MAP.APP_WALLET, label: "app wallet" },
  {
    value: WALLET_MAP.CRYPTO_WALLET,
    label: "crypt wallet",
  },
] as const;

export { WALLET_MAP, WALLET_OPTIONS };
```

ryosuke-sato-51 replied:
@yuki-yamamura 
確かにオブジェクトだとアッパースネークケースで命名しないことの方が多いかもです。。。
すみませんが元のままで大丈夫そうです 🙇

yuki-yamamura replied:
@ryosuke-sato-51 
承知しました 🙇 
今回は元の命名で進めさせていただきます！

ryosuke-sato-51 commented on src/features/app/wallet/components/transfer.tsx:
```diff
@@ -0,0 +1,93 @@
+import { WalletEnum, walletOptions } from "../constants";
+
+export const Transfer = () => {
+  const [selectedWallet] = useState<
```
仕様の認識が間違っていたら申し訳ないのですが、walletのtypeを切り替えるということはベージ遷移することになるので、ここはuseStateにする必要なないかと思いました。

yuki-yamamura replied:
@ryosuke-sato-51 
ありがとうございます 🙇 
画面仕様が理解できておらず、ご指摘の通りstateは不要でした。

ryosuke-sato-51 reviewed (commented):
@yuki-yamamura 
LGTMです！
いくつかコメントしましたので、ご確認よろしくお願いいたします！

yuki-yamamura committed:
fix: フォームのイベントハンドラを修正する

https://github.com/thanklet/mock-dapp-prototype/pull/48#discussion_r1834367372

yuki-yamamura committed:
fix: type="button" を削除する

https://github.com/thanklet/mock-dapp-prototype/pull/48#discussion_r1835606158

yuki-yamamura committed:
refactor: FormControlLabelをラップして使用する

https://github.com/thanklet/mock-dapp-prototype/pull/48#discussion_r1835608434

yuki-yamamura committed:
refactor: 定数の命名を調整する

https://github.com/thanklet/mock-dapp-prototype/pull/48#discussion_r1835609919

yuki-yamamura committed:
fix: 不要なstateを削除する

https://github.com/thanklet/mock-dapp-prototype/pull/48#discussion_r1835610986

yuki-yamamura committed:
Merge branch 'main' into feature/issue-42

yuki-yamamura committed:
feat: Wallet > タブリンクの追加に伴い、ルーティング周りを調整する

情報構造が以下の通りだと仮定してルーティングの整理を行い、crypt-walletのパスを追加した。
- app
  - wallet
    - app-wallet
      - transfer
      - ...
    - crypt-wallet
      - ...

yuki-yamamura committed:
chore: タブリンクのスタイルを調整する

yuki-yamamura committed:
fix: Wallet > Stakingページの表示崩れを修正する

yuki-yamamura committed:
feat: テキストフィールドにロゴを追加する

yuki-yamamura commented:
@ryosuke-sato-51 
レビューありがとうございました！
以下の対応を行いましたので、お手数ですがご確認をいただけますでしょうか。

- レビューの指摘事項を対応
- #50 の対応
- crypt walletの追加を想定して、ルーティングやディレクトリ構造を `app/wallet/**` から `app/wallet/app-wallet/**` と `app/wallet/crypt-wallet/**` に変更

ryosuke-sato-51 reviewed (approved):
@yuki-yamamura 
ご対応ありがとうございます！LGTMです〜
![LGTM](https://firebasestorage.googleapis.com/v0/b/lgtmgen.appspot.com/o/images%2F2afe5b8e-b53f-4f78-b8d2-1087675182d0?alt=media&token=ef1c6bcb-c643-4a52-950e-15452d151319)

yuki-yamamura committed:
chore: 定数名を調整する

yuki-yamamura commented:
@yuki-yamamura 
レビューありがとうございました！
マージします 🙏
